pkg/agent/watch.go

```diff
@@ -262,46 +240,21 @@ func metricsForVM(vm *vmapi.VirtualMachine, nodeName string) map[vmResourceTuple
 	}
 	for _, t := range specMem {
 		if t.value != nil {
-			metrics[vmResourceTuple{
-				namespace:  vm.Namespace,
-				name:       vm.Name,
-				endpointID: endpointID,
-				resource:   vmResourceMem,
-				valueType:  t.key,
-			}] = float64(vm.Spec.Guest.MemorySlotSize.Value() * int64(*t.value))
+			memValue := float64(vm.Spec.Guest.MemorySlotSize.Value() * int64(*t.value))
+			perVMMetrics.memory.WithLabelValues(append(commonLabelValues, string(t.key))...).Set(memValue)
 		}
 	}
-
-	// Status metrics:
-	if vm.Status.CPUs != nil {
-		metrics[vmResourceTuple{
-			namespace:  vm.Namespace,
-			name:       vm.Name,
-			endpointID: endpointID,
-			resource:   vmResourceCPU,
-			valueType:  vmResourceValueStatusUse,
-		}] = vm.Status.CPUs.AsFloat64()
-	}
-	if vm.Status.MemorySize != nil {
-		// nb: convert to int64 first, because (*resource.Quantity).AsApproximateFloat64() is sometimes very inaccurate.
-		metrics[vmResourceTuple{
-			namespace:  vm.Namespace,
-			name:       vm.Name,
-			endpointID: endpointID,
-			resource:   vmResourceMem,
-			valueType:  vmResourceValueStatusUse,
-		}] = float64(vm.Status.MemorySize.Value())
-	}
-
-	return metrics
 }
 
-func (t vmResourceTuple) labelValues() []string {
-	return []string{
-		t.namespace,
-		t.name,
-		t.endpointID,
-		string(t.resource),
-		string(t.valueType),
+func deleteVMMetrics(vm *vmapi.VirtualMachine, nodeName string, perVMMetrics *PerVMMetrics) {
+	if vm.Status.Node != nodeName {
+		return
+	}
+	commonLabels := prometheus.Labels{
+		"vm_namespace": vm.Namespace,
+		"vm_name":      vm.Name,
+		"endpoint_id":  vm.Labels[endpointLabel],
 	}
+	perVMMetrics.cpu.DeletePartialMatch(commonLabels)
+	perVMMetrics.memory.DeletePartialMatch(commonLabels)
```

Comment:
Not sure how I feel about using label names here when we're using positional labels everywhere else. I'd be worried about changing the labels in the metric but forgetting to change it here.

Otherwise, seems like a clean solution.

Reply:
what do you mean by "positional labels"?

Reply:
I suppose just that those `WithLabelValues` -methods take vaues as positional arguments, but the delete takes labels as key-value pairs (and doesn’t therefore care about order for example). That’s a property of the library though.

It’s not mandatory to use partial, just thought it might simplify a bit.